ark/phenotyping/cluster_helpers.py

```diff
@@ -1,16 +1,310 @@
+from abc import ABC, abstractmethod
 import bisect
 import feather
 from itertools import combinations
 import numpy as np
+import os
 import pathlib
 import pandas as pd
+from pyFlowSOM import map_data_to_nodes, som
 from scipy.stats import zscore
 from sklearn.cluster import AgglomerativeClustering
 from typing import Callable, List, Protocol, runtime_checkable
+import warnings
 
+from ark.utils.io_utils import list_files, validate_paths
 from ark.utils.misc_utils import verify_in_list
 
 
+class PixieSOMCluster(ABC):
+    @abstractmethod
+    def __init__(self, weights_path: pathlib.Path, columns: List[str], num_passes: int = 1,
+                 xdim: int = 10, ydim: int = 10, lr_start: float = 0.05, lr_end: float = 0.01):
+        """Generic implementation of a pyFlowSOM runner
+
+        Args:
+            weights_path (pathlib.Path):
+                The path to save the weights to.
+            columns (List[str]):
+                The list of columns to subset the data on.
+            num_passes (int):
+                The number of SOM training passes to use.
+            xdim (int):
+                The number of SOM nodes on the x-axis.
+            ydim (int):
+                The number of SOM nodes on the y-axis.
+            lr_start (float):
+                The initial learning rate.
+            lr_end (float):
+                The learning rate to decay to
+        """
+        self.weights_path = weights_path
+        self.weights = None if not os.path.exists(weights_path) else feather.read_dataframe(
+            weights_path
+        )
+        self.columns = columns
+        self.num_passes = num_passes
+        self.xdim = xdim
+        self.ydim = ydim
+        self.lr_start = lr_start
+        self.lr_end = lr_end
+
+    @abstractmethod
+    def normalize_data(self) -> pd.DataFrame:
+        """Generic implementation of the normalization process to use on the input data
+
+        Returns:
+            pandas.DataFrame:
+                The data with `columns` normalized by the values in `norm_data`
+        """
+        pass
+
+    def train_som(self, data: pd.DataFrame):
+        """Trains the SOM on the data provided and saves the weights generated
+
+        Args:
+            data (pandas.DataFrame):
+                The input data to train the SOM on.
+        """
+        # make sure to run a deterministic SOM for reproducibility purposes
+        som_weights = som(
```

Comment:
In the `som` function (and correspondingly in init), I think it would be good to have the `rlen` parameter (the number of passes through the dataset - that is something that might be changed).